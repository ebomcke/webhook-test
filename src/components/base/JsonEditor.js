import React from 'react';
import brace from 'brace';
import 'brace/mode/json';
import 'brace/theme/github';
import AceEditor from 'react-ace';

const JsonEditor = ({ height, value, onChange }) => (
  <AceEditor
    mode="json"
    theme="github"
    wrapEnabled={true}
    width="100%"
    height={height}
    value={value}
    onChange={onChange}
  />
);

export default JsonEditor;