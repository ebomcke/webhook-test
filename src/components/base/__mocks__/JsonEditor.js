import React from 'react';

const JsonEditor = props => (
  <div>
    This is a placeholder for the JSON editor.
    <pre>
      <code>{JSON.stringify(props)}</code>
    </pre>
  </div>
);

export default JsonEditor;