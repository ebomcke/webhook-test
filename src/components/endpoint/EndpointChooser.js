import React from 'react';
import { compose } from 'recompose';
import { withAccount } from '../../contexts/Account';
import { Dropdown } from 'semantic-ui-react';
import { withEndpoints } from '../../contexts/Endpoints';

const EndpointChooser = ({ account, endpoints, onChange }) => (
  <Dropdown
    selection
    search
    options={endpoints.map(endpoint => ({
      text: `/${account.organisationName}/${endpoint.path}`,
      value: endpoint.id,
    }))}
    onChange={(event, data) =>
      onChange(endpoints.find(e => e.id === data.value))
    }
  />
);
export default compose(
  withAccount,
  withEndpoints,
)(EndpointChooser);
