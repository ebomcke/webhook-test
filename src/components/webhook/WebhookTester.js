import React, { Component } from 'react';
import { Form, Button } from 'semantic-ui-react';
import { compose } from 'recompose';
import { withAccount } from '../../contexts/Account';
import JsonEditor from '../base/JsonEditor';
import EndpointChooser from '../endpoint/EndpointChooser';

class WebhookTester extends Component {
  constructor(props) {
    super(props);
    this.state = {
      body: JSON.stringify(
        {
          status: 'OK',
        },
        null,
        2,
      ),
      endpoint: props.endpoint,
    };
    this.sendMessage = this.sendMessage.bind(this);
    this.handleEndpointSelect = this.handleEndpointSelect.bind(this);
    this.handleJsonInputChange = this.handleJsonInputChange.bind(this);
  }

  async sendMessage() {
    const { account, onComplete } = this.props;
    this.setState({ isLoading: true });
    try {
      const response = await fetch(
        `${process.env.API_URL}/webhook/${account.organisationName}/${
          this.state.endpoint.path
        }`,
        {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(JSON.parse(this.state.body)),
        },
      );
      if (response.status !== this.state.endpoint.defaultResponse.statusCode) {
        this.setState({ isLoading: false, isError: true });
      } else {
        onComplete && onComplete();
      }
    } catch (e) {
      console.log(e);
      this.setState({ isLoading: false, isError: true });
    }
  }

  handleEndpointSelect(endpoint) {
    this.setState({
      endpoint: endpoint,
    });
  }

  handleJsonInputChange(value) {
    try {
      JSON.parse(value);
      this.setState({
        body: value,
      });
    } catch (e) {}
  }

  render() {
    const { endpoint } = this.props;
    return (
      <Form>
        {!endpoint && (
          <Form.Field required>
            <label>Url</label>
            <EndpointChooser onChange={this.handleEndpointSelect} />
          </Form.Field>
        )}
        <Form.Field>
          <label>Body</label>
          <JsonEditor
            height="300px"
            value={this.state.body}
            onChange={this.handleJsonInputChange}
          />
        </Form.Field>
        <Button
          onClick={this.sendMessage}
          loading={this.state.isLoading}
          disabled={!this.state.endpoint}
        >
          Send
        </Button>
      </Form>
    );
  }
}

export default compose(withAccount)(WebhookTester);
