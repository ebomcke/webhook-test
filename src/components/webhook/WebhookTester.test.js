import React from 'react';
import { create } from 'react-test-renderer';
import WebhookTester from './WebhookTester';
import EndpointChooser from '../endpoint/EndpointChooser';
import JsonEditor from '../base/JsonEditor';

describe('Rendering tests', () => {
  test('No endpoint provided - dropdown rendered', () => {
    const webhookTester = create(<WebhookTester />);
    expect(webhookTester.toJSON()).toMatchSnapshot();
  });

  test('Endpoint provided - no dropdown rendered', () => {
    const endpoint = {
      path: 'unit-test',
    };
    const webhookTester = create(<WebhookTester endpoint={endpoint} />);
    expect(webhookTester.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  let webhookTester;
  beforeEach(() => {
    webhookTester = create(<WebhookTester />);
  });
  test('Selecting endpoint updates the state', () => {
    const endpointChooser = webhookTester.root.findByType(EndpointChooser);
    const endpoint = {
      path: 'selected-endpoint',
    };
    endpointChooser.props.onChange(endpoint);
    const instance = webhookTester.root.children[0].instance;
    expect(instance.state.endpoint).toEqual(endpoint);
  });

  test('Updating JSON body updates the state', () => {
    const jsonEditor = webhookTester.root.findByType(JsonEditor);
      const newBody = JSON.stringify({
          updated: 'value',
      });
    jsonEditor.props.onChange(
      newBody,
    );
    const instance = webhookTester.root.children[0].instance;
    expect(instance.state.body).toEqual(newBody);
  });

  test('Updating JSON body does not update the state if json is invalid', () => {
    const jsonEditor = webhookTester.root.findByType(JsonEditor);
    const instance = webhookTester.root.children[0].instance;
    const body = instance.state.body;
    jsonEditor.props.onChange('{ "invalid": json');
    expect(instance.state.body).toEqual(body);
  });
});
