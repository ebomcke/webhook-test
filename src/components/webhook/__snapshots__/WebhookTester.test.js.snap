// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests Endpoint provided - no dropdown rendered 1`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label>
      Body
    </label>
    <div>
      This is a placeholder for the JSON editor.
      <pre>
        <code>
          {"height":"300px","value":"{\\n  \\"status\\": \\"OK\\"\\n}"}
        </code>
      </pre>
    </div>
  </div>
  <button
    className="ui button"
    onClick={[Function]}
  >
    Send
  </button>
</form>
`;

exports[`Rendering tests No endpoint provided - dropdown rendered 1`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="required field"
  >
    <label>
      Url
    </label>
    <div
      aria-expanded={false}
      className="ui search selection dropdown"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        autoComplete="off"
        className="search"
        onChange={[Function]}
        style={
          Object {
            "width": undefined,
          }
        }
        tabIndex={0}
        type="text"
        value=""
      />
      <div
        aria-live="polite"
        className="text"
        role="alert"
      />
      <i
        aria-hidden="true"
        className="dropdown icon"
        onClick={[Function]}
      />
      <div
        className="menu transition"
        role="listbox"
      >
        <div
          aria-checked={false}
          aria-selected={true}
          className="selected item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            /unit-test-org/unit-test-1
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            /unit-test-org/unit-test-2
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="field"
  >
    <label>
      Body
    </label>
    <div>
      This is a placeholder for the JSON editor.
      <pre>
        <code>
          {"height":"300px","value":"{\\n  \\"status\\": \\"OK\\"\\n}"}
        </code>
      </pre>
    </div>
  </div>
  <button
    className="ui disabled button"
    disabled={true}
    onClick={[Function]}
    tabIndex={-1}
  >
    Send
  </button>
</form>
`;
